std/digest/murmurhash.d

```diff
@@ -500,28 +509,75 @@ struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 6
         // Buffer should never be full while entering this function.
         assert(bufferSize < Element.sizeof);
 
-        // Check if we have some leftover data in the buffer. Then fill the first block buffer.
+        // Check if we don't even fill up a single block buffer.
         if (bufferSize + data.length < Element.sizeof)
         {
             buffer.data[bufferSize .. bufferSize + data.length] = data[];
             bufferSize += data.length;
             return;
         }
-        const bufferLeeway = Element.sizeof - bufferSize;
-        assert(bufferLeeway <= Element.sizeof);
-        buffer.data[bufferSize .. $] = data[0 .. bufferLeeway];
-        putElement(buffer.block);
-        data = data[bufferLeeway .. $];
+
+        // Check if we have some leftover data in the buffer. Then fill the first block buffer.
+        const bool haveLeewayElement = (bufferSize != 0);
+        if (haveLeewayElement)
+        {
+            const bufferLeeway = Element.sizeof - bufferSize;
+            assert(bufferLeeway < Element.sizeof);
+            buffer.data[bufferSize .. $] = data[0 .. bufferLeeway];
+            putElement(buffer.block);
+            element_count += Element.sizeof;
+            data = data[bufferLeeway .. $];
+        }
 
         // Do main work: process chunks of `Element.sizeof` bytes.
         const numElements = data.length / Element.sizeof;
         const remainderStart = numElements * Element.sizeof;
-        foreach (ref const Element block; cast(const(Element[]))(data[0 .. remainderStart]))
+        version (haveUnalignedLoads)
```

Comment:
Haven't we already talked about this before?  Or am I thinking of another patch.

Reply:
Possibly, not with me though; a first incarnation of this murmurhash patch is in https://github.com/dlang/phobos/pull/5358 (which I only realized after creating this PR). This one here still isn't fully satisfactory as casting a static array to another static array type may not be well-defined in the spec (haven't checked, but it's reinterpret_cast vs. conversion to new rvalue). The latest (LDC-specific) version without ambiguities is https://github.com/ldc-developers/phobos/pull/44.

Reply:
Now in PascalCase.